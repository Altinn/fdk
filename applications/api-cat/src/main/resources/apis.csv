OrgNr;OrgName;ApiSpecUrl;DocUrl;AccessRights;Provenance;DatasetRefs
974760673;Brønnøysundregistrene;;https://data.brreg.no/enhetsregisteret/api/docs/index.html;NON_PUBLIC,RESTRICTED;NASJONAL;https://fellesdatakatalog.brreg.no/datasets/ae335ad2-0f3d-4eb1-98ce-87882776a123,https://fellesdatakatalog.brreg.no/datasets/ae335ad2-0f3d-4eb1-98ce-87882776a124
974760673;Brønnøysundregistrene;https://fellesdatakatalog.brreg.no/api-docs/search-api;https://fellesdatakatalog.brreg.no/swagger-ui.html;NON_PUBLIC,RESTRICTED;NASJONAL;https://fellesdatakatalog.brreg.no/datasets/ea9e2aa7-880c-4387-888e-77446acab279,https://fellesdatakatalog.brreg.no/datasets/ea9e2aa7-880c-4387-888e-77446acab278
974760673;;https://docker-demo.fsat.no/fsapi/swagger.json;https://docker-demo.fsat.no/fsapi/swagger.html#!/person/get_personer_id;;;
974760673;;https://ks-no.github.io/api/innsyn-sok-api-v1.json;https://editor.swagger.io/?url=https://ks-no.github.io/api/innsyn-sok-api-v1.json;;;https://data.norge.no/node/78,https://data.norge.no/node/78
974760673;;https://oidc-ver2.difi.no/kontaktinfo-oauth2-server/rest/apis;https://oidc-ver2.difi.no/kontaktinfo-oauth2-server/rest/swagger-ui.html;;;https://data.norge.no/node/78