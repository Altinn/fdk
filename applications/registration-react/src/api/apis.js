import axios from 'axios';
import qs from 'qs';

export const getAPIByTitlePrefix = async (
  title = '',
  orgPath,
  returnFields
) => {
  const queryParams = `title=${title}${orgPath ? `&orgPath=${orgPath}` : ''}${
    returnFields ? `&returnfields=${returnFields}` : ''
  }`;

  const url = `/api/apis?${queryParams}`;

  const response = await axios
    .get(url)
    .catch(e => console.log(JSON.stringify(e))); // eslint-disable-line no-console

  return response && response.data;
};

export const getApiByDatasetUri = async (datasetUri, returnFields) => {
  const query = `${qs.stringify(
    {
      dataseturi: datasetUri,
      returnFields
    },
    { addQueryPrefix: true }
  )}`;

  const url = `/api/apis${query}`;

  const response = await axios
    .get(url)
    .catch(e => console.log(JSON.stringify(e))); // eslint-disable-line no-console

  return response && response.data;
};
