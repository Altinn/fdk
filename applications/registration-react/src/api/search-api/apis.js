import { getConfig } from '../../config';
import { searchApiGet } from './host';

export const searchApis = params =>
  searchApiGet({
    url: `${getConfig().apiApi.host}/api/apis`,
    params,
    authorization: getConfig().apiApi.authorization
  });
