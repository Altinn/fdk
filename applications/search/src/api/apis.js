import qs from 'qs';
import axios from 'axios';

const apisUrlBase = '/api/apis';

export const apisSearchUrl = query =>
  `${apisUrlBase}/search${qs.stringify(query, { addQueryPrefix: true })}`;

export const getApi = id =>
  axios
    .get(`${apisUrlBase}/${id}`)
    .then(response => response.data)
    .catch(e => console.error(JSON.stringify(e))); // eslint-disable-line no-console;
