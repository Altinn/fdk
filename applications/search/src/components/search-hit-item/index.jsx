import * as React from "react";
import PropTypes from 'prop-types';
import * as _ from "lodash";
import cx from 'classnames';
const qs = require('qs');
const defaults = require("lodash/defaults");

import localization from '../../components/localization';
import './index.scss';

export default class SearchHitItem extends React.Component { // eslint-disable-line react/prefer-stateless-function
  constructor(props) {
    super(props);
    this.state = {
      result: props.result
    };
    this.extractDomain = this.extractDomain.bind(this);
  }

  componentDidUpdate() {
    console.log("search-hit-item componentDidUpdate");
    this.state = {
      result: this.props.result
    };
  }

  extractDomain(url) {
    var domain;
    //find & remove protocol (http, ftp, etc.) and get domain
    if (url.indexOf("://") > -1) {
      domain = url.split('/')[2];
    }
    else {
      domain = url.split('/')[0];
    }
    //find & remove port number
    domain = domain.split(':')[0];
    return domain;
  }

  _renderFormats(source) {
    let distribution = source.distribution;
    let formatRet;
    if(source.distribution) {
      formatRet = Object.keys(distribution).map(key => {
        if(distribution[key].format) {
          let formatArray = distribution[key].format.trim().split(',');
          let ret = '';
          formatArray.forEach((format) => {
            if (format !== null) {
              ret += '<div class="fdk-button-format fdk-button-format-inactive">';
              ret += '<i class="fa fa-download fdk-fa-left"></i>';
              ret += format;
              ret += '</div>';
            }
          });
          return ret;
        } return null;
      });
    }
    return formatRet;
  }

  render() {
    const result = this.state.result;
    let url =  'datasets?id=' + encodeURIComponent(result._id);
    //let queryObj = qs.parse(window.location.search.substr(1));
    let queryObj = qs.parse(window.location.search);
    let language = queryObj.lang ? queryObj.lang : 'nb';
    if(!queryObj.lang || queryObj.lang === 'nb') {
      url += '&lang=nb';
    } else if(queryObj.lang === 'nn'){
      url += '&lang=nn';
    } else {
      url += '&lang=en';
    }
    const source = _.extend({}, result._source, result.highlight);
    const hit_id = encodeURIComponent(source.id);
    const hit_element_id = `search-hit-${hit_id}`;
    const title = source.title[language] ? source.title[language] : '';
    const link = `/datasets?id=${hit_id}`;

    let themeLabels = '';
    if(source.theme) {
      source.theme.forEach((singleTheme, index)=> {
        if(singleTheme.title) {
          themeLabels += '<div class="fdk-label fdk-label-on-white">';
          themeLabels += singleTheme.title[language] || singleTheme.title.nb || singleTheme.title.nn || singleTheme.title.en;
          themeLabels += ' </div>';
        }
      });
    }

    let distribution_restricted = false;
    let distribution_non_public = false;
    let distribution_public = false;

    if (source.accessRights && source.accessRights.authorityCode === 'RESTRICTED') {
      distribution_restricted = true;
    } else if (source.accessRights && source.accessRights.authorityCode === 'PUBLIC') {
      distribution_public = true;
    } else if (source.accessRights && source.accessRights.authorityCode === 'NON_PUBLIC') {
      distribution_non_public = true;
    } else if (!source.accessRights) { // antar public hvis authoritycode mangler
      distribution_public = true;
    }

    const distributionClass = cx(
      'fdk-container-distributions',
      {
        'fdk-distributions-red': distribution_restricted,
        'fdk-distributions-yellow': distribution_non_public,
        'fdk-distributions-green': distribution_public
      }
    );

    return (
      <a
        id={hit_element_id}
        className="fdk-a-search-hit"
        title={`${localization.result.dataset}: ${title}`}
        href={link}
        rel="noopener noreferrer"
      >
        <div className="fdk-container fdk-container-search-hit">
        <h2>{title}</h2>
        <div>
          {localization.search_hit.owned} <span href="#">{source.publisher ? source.publisher.name.charAt(0) + source.publisher.name.substring(1).toLowerCase() : ''}</span>
          <span dangerouslySetInnerHTML={{__html:themeLabels}}></span>
        </div>
        <p
          className="fdk-p-search-hit"
          dangerouslySetInnerHTML={
            {__html:source.description[language] || source.description.nb || source.description.nn || source.description.en}}
        >
        </p>
        <div className={distributionClass}>
          <strong>{source.accessRights ? source.accessRights.prefLabel[language] : 'Offentlig'}</strong>
          <br />
          <div dangerouslySetInnerHTML={{__html: this._renderFormats(source)}} />
        </div>
        {source.landingPage ? <div dangerouslySetInnerHTML={{__html:landingPage}}/> : ''}
        </div>
      </a>
    );
  }
}

SearchHitItem.defaultProps = {
  result: null
};

SearchHitItem.propTypes = {
  result: PropTypes.object
};
