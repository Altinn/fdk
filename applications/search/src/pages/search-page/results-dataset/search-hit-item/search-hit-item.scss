@import '../../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.fdk-a-search-hit {
  text-decoration: none;
}

.fdk-themes-text {
  padding-top: 0.5em;
}

.theme-fdk {
  .fdk-icon-public {
    background-image: url("/static/img/icon-dataset-public-md.svg");
  }
}

.theme-nap {
  .fdk-icon-public {
    background-image: url("/static/img/icon-dataset-public-nap-md.svg");
  }
}

.fdk-icon-public {
  background-image: url("/static/img/icon-dataset-public-md.svg");
  width: 2em; 
  height: 2em;
}

.fdk-icon-non-public {
  background-image: url("/static/img/icon-dataset-not-public-md.svg");
  width: 2em; 
  height: 2em;
}

.fdk-icon-restricted {
  background-image: url("/static/img/icon-dataset-restricted-md.svg");
  width: 2em; 
  height: 2em;
}

.fdk-color-restricted {
  color: $color-warning-darker !important;
}