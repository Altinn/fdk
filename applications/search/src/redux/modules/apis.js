import _ from 'lodash';
import { fetchActions } from '../fetchActions';

export const APIS_REQUEST = 'APIS_REQUEST';
export const APIS_SUCCESS = 'APIS_SUCCESS';
export const APIS_FAILURE = 'APIS_FAILURE';

function shouldFetch(metaState, query) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      ((metaState.lastFetch || 0) < Date.now() - threshold ||
        metaState.cachedQuery !== query))
  );
}

export function fetchApisIfNeededAction(query) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['apis', 'meta']), query) &&
    dispatch(
      fetchActions(`/api/apis/search?${query}`, [
        { type: APIS_REQUEST, meta: { query } },
        { type: APIS_SUCCESS, meta: { query } },
        APIS_FAILURE
      ])
    );
}

const initialState = {};

export function apis(state = initialState, action) {
  switch (action.type) {
    case APIS_REQUEST:
      return {
        ...state,
        meta: {
          isFetching: true,
          lastFetch: null,
          cachedQuery: action.meta.query
        }
      };
    case APIS_SUCCESS:
      return {
        ...state,
        apiItems: _.get(action.payload, 'hits'),
        apiAggregations: _.get(action.payload, 'aggregations'),
        apiTotal: _.get(action.payload, 'total'),
        meta: {
          isFetching: false,
          lastFetch: Date.now(),
          cachedQuery: action.meta.query
        }
      };
    case APIS_FAILURE:
      return {
        ...state,
        apiItems: null,
        apiAggregations: null,
        apiTotal: null,
        meta: {
          isFetching: false,
          lastFetch: null,
          cachedQuery: null
        }
      };
    default:
      return state;
  }
}
