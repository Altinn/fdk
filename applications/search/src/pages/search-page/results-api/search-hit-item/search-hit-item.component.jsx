import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import cx from 'classnames';
import _ from 'lodash';

import localization from '../../../../lib/localization';
import './search-hit-item.scss';
import { getTranslateText } from '../../../../lib/translateText';
import { SearchHitHeader } from '../../../../components/search-hit-header/search-hit-header.component';
import { convertToSanitizedHtml } from '../../../../lib/markdown-converter';

const renderHeaderLink = (item, publisher, publishers, referenceData) => {
  if (!item) {
    return null;
  }
  const { title } = item;
  const link = `/apis/${encodeURIComponent(item.id)}`;

  return (
    <header>
      <SearchHitHeader
        tag="h2"
        title={getTranslateText(title)}
        titleLink={link}
        publisherLabel={localization.api.provider}
        publisher={publisher}
        publisherItems={publishers}
        nationalComponent={item.nationalComponent}
        statusCode={item.statusCode}
        referenceData={referenceData}
      />
    </header>
  );
};

const renderDescription = descriptionFormatted => {
  if (!descriptionFormatted) {
    return null;
  }

  const sanitizedHtml = convertToSanitizedHtml(descriptionFormatted);

  // our naive cropping algorithm brutally abrupts the markup string.
  // It is a bit safer to do it in html markup than in markdown.
  const croppedHtml =
    sanitizedHtml.length < 250
      ? sanitizedHtml
      : `${sanitizedHtml.substr(0, 220)}...`;

  return (
    <div className="fdk-text-size-medium">
      <div className="uu-invisible" aria-hidden="false">
        Beskrivelse av api
      </div>
      <div
        dangerouslySetInnerHTML={{
          __html: croppedHtml
        }}
      />
    </div>
  );
};

const renderExpiredVersion = expired => {
  if (!expired) {
    return null;
  }
  return (
    <div className="search-hit__version mb-4 p-4">
      <span>Denne versjonen av API-et er utgått og vil fases ut i 2019. </span>
      <Link to="/TODO">Versjon 2 er dokumentert her.</Link>
    </div>
  );
};

const renderAccessRights = accessRights => {
  if (!(accessRights && accessRights[0])) {
    return null;
  }

  const { code } = accessRights[0] || {
    code: null
  };
  let accessRightsLabel;

  switch (code) {
    case 'NON_PUBLIC':
      accessRightsLabel = localization.api.accessRight.nonPublic;
      break;
    case 'RESTRICTED':
      accessRightsLabel = localization.api.accessRight.restricted;
      break;
    case 'PUBLIC':
      accessRightsLabel = localization.api.accessRight.public;
      break;
    default:
      accessRightsLabel = localization.api.accessRight.unknown;
  }

  return (
    <div className="mb-2">
      <strong>{accessRightsLabel}</strong>
    </div>
  );
};

const renderFormat = formats => {
  if (!formats || formats.length === 0) {
    return null;
  }

  const formatItems = formats =>
    formats.map((item, index) => (
      <span key={index}>
        {index > 0 ? ', ' : ''}
        {item}
      </span>
    ));

  return (
    <div className="mb-2">
      <strong>{localization.format}:&nbsp;</strong>
      {formatItems(formats)}
    </div>
  );
};

export const SearchHitItem = ({
  item,
  fadeInCounter,
  publishers,
  referenceData
}) => {
  const searchHitClass = cx('search-hit', {
    'fade-in-200': fadeInCounter === 0,
    'fade-in-300': fadeInCounter === 1,
    'fade-in-400': fadeInCounter === 2
  });

  return (
    <article className={searchHitClass}>
      <span className="uu-invisible" aria-hidden="false">
        Søketreff.
      </span>

      {renderHeaderLink(
        item,
        _.get(item, 'publisher'),
        publishers,
        referenceData
      )}

      {renderExpiredVersion(_.get(item, 'expired'))}

      {renderDescription(
        _.get(_.get(item, 'descriptionFormatted')) || _.get(item, 'description')
      )}

      {renderAccessRights(_.get(item, 'accessRights'))}

      {renderFormat(_.get(item, 'formats'))}
    </article>
  );
};

SearchHitItem.defaultProps = {
  fadeInCounter: null,
  item: null,
  publishers: null,
  referenceData: null
};

SearchHitItem.propTypes = {
  fadeInCounter: PropTypes.number,
  item: PropTypes.shape({}),
  publishers: PropTypes.object,
  referenceData: PropTypes.object
};
