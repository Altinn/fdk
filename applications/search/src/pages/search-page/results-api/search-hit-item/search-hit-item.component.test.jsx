import React from 'react';
import { shallow } from 'enzyme';
import { SearchHitItem } from './search-hit-item.component';
import apis from '../../../../../test/fixtures/apis';

let defaultProps;

test('should render api-SearchHitItem correctly', () => {
  defaultProps = {
    result: apis[0].hits[0]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
