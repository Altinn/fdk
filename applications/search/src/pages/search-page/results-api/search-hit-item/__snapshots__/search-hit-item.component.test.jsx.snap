// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render api-SearchHitItem correctly 1`] = `
<article
  className="search-hit"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
  <header>
    <Link
      className="search-hit__title-link"
      replace={false}
      title="[object Object]: National Data Directory Search API"
      to="/apis/jar%3Afile%3A%2Fapp.jar!%2FBOOT-INF%2Fclasses!%2Fdatakatalog-api.json"
    >
      <SearchHitHeader
        headerTag="h2"
        provenance={null}
        publisher={null}
        publisherItems={null}
        publisherLabel="Provider"
        theme={null}
        title="National Data Directory Search API"
      />
    </Link>
  </header>
  <p>
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Beskrivelse av api,
    </span>
    Provides a basic search api against the National Data Directory of Norway
  </p>
</article>
`;

exports[`should render api-SearchHitItem correctly with no props 1`] = `
<article
  className="search-hit"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
</article>
`;

exports[`should render api-SearchHitItem with long description 1`] = `
<article
  className="search-hit"
>
  <span
    aria-hidden="false"
    className="uu-invisible"
  >
    Søketreff.
  </span>
  <header>
    <Link
      className="search-hit__title-link"
      replace={false}
      title="[object Object]: National Data Directory Search API"
      to="/apis/jar%3Afile%3A%2Fapp.jar!%2FBOOT-INF%2Fclasses!%2Fdatakatalog-api.json"
    >
      <SearchHitHeader
        headerTag="h2"
        provenance={null}
        publisher={null}
        publisherItems={null}
        publisherLabel="Provider"
        theme={null}
        title="National Data Directory Search API"
      />
    </Link>
  </header>
  <p>
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Beskrivelse av api,
    </span>
    It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as...
  </p>
</article>
`;
